### PR TITLE
break: upgrade to typescript 5.2, typedoc 0.25

### PR DESCRIPTION
**Pull request by adidahiya**
_This change will enable us to upgrade Blueprint to TypeScript v5.2 in https://github.com/palantir/blueprint/pull/6476_

✅  Prerequisites:

- [x] https://github.com/palantir/documentalist/pull/236

Changes proposed:

- Upgrade to TypeScript v5.2
- Upgrade to typedoc v0.25
- Migrate to latest typedoc APIs (see [changelog](https://typedoc.org/guides/changelog/))
- Upgrade Jest & ts-jest to v29
- Update snapshots for changes related to Jest syntax breaks (the redundant `Object`, `Array` prefixes on literals are no longer there)
- Update snapshots for typedoc changes
- ~Add [typedoc-plugin-missing-exports](https://www.npmjs.com/package/typedoc-plugin-missing-exports) plugin to bring back some functionality that existed before typedoc v0.20~ edit: I will consider doing this in a separate PR
- Update the TypeScript plugin to handle documentation of files from different TypeScript projects, something that came "for free" before TypeDoc rehauled its file/program resolution system in v0.20.
  - This is accomplished by attempting to resolve each input file with its closest tsconfig.json project config using [tsconfig-resolver](https://www.npmjs.com/package/tsconfig-resolver)

References
- https://typedoc.org/guides/installation/
- https://github.com/palantir/documentalist/pull/144
- https://github.com/TypeStrong/typedoc/commit/50c2fe0fd72055c58a132d22e5a9846d3850a06f
- https://github.com/TypeStrong/typedoc/blob/master/CHANGELOG.md

### CODE REVIEW

File: tslint.json [hunk 1]
```diff
@@ -11,6 +11,8 @@
         "Copyright 2022 Palantir Technologies, Inc. All rights reserved.\n\nLicensed under the Apache License, Version 2.0 (the \"License\");\nyou may not use this file except in compliance with the License.\nYou may obtain a copy of the License at\n\n    http://www.apache.org/licenses/LICENSE-2.0\n\nUnless required by applicable law or agreed to in writing, software\ndistributed under the License is distributed on an \"AS IS\" BASIS,\nWITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\nSee the License for the specific language governing permissions and\nlimitations under the License."
       ]
     },
+    "no-invalid-this": false,
+    "no-shadowed-variable": false,
```
Comment (adidahiya) [inline tslint.json:15]: these lint rules are not compatible with TypeScript v5.x since they use deprecated APIs 🤷🏽 I'll migrate to eslint in a separate PR

File: scripts/submit-comment-with-artifact-links.mjs [hunk 2]
```diff
@@ -0,0 +1,84 @@
+#!/usr/bin/env node
```
Comment (adidahiya) [inline scripts/submit-comment-with-artifact-links.mjs]: this script was copied from the Blueprint repo

File: scripts/submit-preview-comment.sh [hunk 3]
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env bash
```
Comment (adidahiya) [inline scripts/submit-preview-comment.sh]: this script was copied from the Blueprint repo

File: .circleci/config.yml [hunk 4]
```diff
@@ -3,7 +3,7 @@ version: 2.1
 references:
   setup_env: &setup_env
     docker:
-      - image: cimg/node:16.14
+      - image: cimg/node:18.18
```
Comment (adidahiya) [inline .circleci/config.yml]: I could potentially extract this upgrade into a separate PR, hopefully there are no type conflicts with `lib.d.ts` in the current version of TypeScript

File: .circleci/config.yml [hunk 5]
```diff
@@ -76,12 +76,7 @@ jobs:
       - restore_cache: *restore_cache
       - attach_workspace: { at: "." }
       - store_artifacts: { path: packages/docs/dist }
-      - run:
-          name: Submit Github comment with links to built artifacts
-          command: |
-            artifacts=$(curl -X GET "https://circleci.com/api/v2/project/github/palantir/documentalist/$CIRCLE_BUILD_NUM/artifacts" -H "Accept: application/json" -u "$CIRCLE_AUTH_TOKEN:")
-            echo $artifacts > ./scripts/artifacts.json
-            node ./scripts/circle-build-preview.js
+      - run: ./scripts/submit-preview-comment.sh
```
Comment (adidahiya) [inline .circleci/config.yml]: I could extract this to a separate PR - this CI job is failing on develop right now

**Comment by adidahiya** [comment]
the preview comment is not working since we don't have the API token being injected right now but we do have docs built and saved as an artifact: https://output.circle-artifacts.com/output/job/754a7f7f-96d9-43c1-9910-10f2e06a60af/artifacts/0/packages/docs/dist/index.html

File: packages/compiler/src/__tests__/__snapshots__/typescript.test.ts.snap [hunk 6]
```diff
@@ -425,196 +407,18 @@ Set the noise for the animal",
       },
     ],
     "name": "Dog",
-    "properties": Array [],
+    "properties": [],
     "sourceUrl": "https://github.com/palantir/documentalist/blob/develop/packages/compiler/src/__tests__/__fixtures__/classes.ts#L53",
   },
-  "Food": Object {
```
Comment (adidahiya) [inline packages/compiler/src/__tests__/__snapshots__/typescript.test.ts.snap:431]: non-exported symbols are no longer documented. I think that's OK.

File: packages/compiler/src/__tests__/__snapshots__/typescript.test.ts.snap [hunk 7]
```diff
@@ -221,24 +212,15 @@ Set the noise for the animal",
...
-",
-          ],
-          "contentsRaw": "Get the noise of the animal
-Set the noise for the animal",
-          "metadata": Object {},
-        },
+  "Dog": {
+    "accessors": [
+      {
+        "documentation": undefined,
```
Comment (adidahiya) [inline packages/compiler/src/__tests__/__snapshots__/typescript.test.ts.snap:221]: need to look into why `get` accessors are no longer getting documented. this is low-priority, though, since we rarely use these in our frontend code bases.

File: packages/compiler/src/__tests__/__snapshots__/typescript.test.ts.snap [hunk 8]
```diff
@@ -650,12 +454,12 @@ Object {
       "isStatic": false,
     },
     "kind": "enum",
-    "members": Array [
-      Object {
-        "defaultValue": "\\"danger\\"",
+    "members": [
+      {
+        "defaultValue": undefined,
```
Comment (adidahiya) [inline packages/compiler/src/__tests__/__snapshots__/typescript.test.ts.snap]: default values are getting dropped - not sure why 🤔

File: packages/compiler/src/__tests__/__snapshots__/typescript.test.ts.snap [hunk 9]
```diff
@@ -747,34 +551,34 @@ Object {
     },
     "inheritedFrom": undefined,
     "kind": "method",
-    "name": "numberAndString",
-    "signatures": Array [
-      Object {
-        "documentation": Object {
-          "contents": Array [
-            "<p>Exported function.</p>
+    "name": "MyFunc",
```
Comment (adidahiya) [inline packages/compiler/src/__tests__/__snapshots__/typescript.test.ts.snap:554]: these functions are still documented correctly, but their order got moved around in the snapshot

File: packages/compiler/src/__tests__/__snapshots__/typescript.test.ts.snap [hunk 10]
```diff
@@ -993,18 +801,11 @@ Object {
...
         "defaultValue": undefined,
-        "documentation": Object {
-          "contents": Array [
-            "<p>Click event handler.</p>
-",
-          ],
-          "contentsRaw": "Click event handler.",
-          "metadata": Object {},
-        },
+        "documentation": undefined,
```
Comment (adidahiya) [inline packages/compiler/src/__tests__/__snapshots__/typescript.test.ts.snap]: this jsdoc got dropped - not sure why 🤔

File: packages/compiler/src/__tests__/__snapshots__/typescript.test.ts.snap [hunk 11]
```diff
@@ -1067,25 +868,18 @@ Object {
...
-      "documentation": Object {
-        "contents": Array [
-          "<p>Index signature for the masses.</p>
-",
-        ],
-        "contentsRaw": "Index signature for the masses.",
-        "metadata": Object {},
-      },
+    "indexSignature": {
+      "documentation": undefined,
```
Comment (adidahiya) [inline packages/compiler/src/__tests__/__snapshots__/typescript.test.ts.snap:879]: this index signature jsdoc got dropped - not sure why 🤔

File: packages/compiler/src/__tests__/__snapshots__/typescript.test.ts.snap [hunk 12]
```diff
@@ -1165,18 +969,11 @@ This is equivalent to setting \`className=\\"pt-active\\"\`.",
...
         "defaultValue": undefined,
-        "documentation": Object {
-          "contents": Array [
-            "<p>A ref handler that receives the native HTML element backing this component.</p>
-",
-          ],
-          "contentsRaw": "A ref handler that receives the native HTML element backing this component.",
-          "metadata": Object {},
-        },
+        "documentation": undefined,
```
Comment (adidahiya) [inline packages/compiler/src/__tests__/__snapshots__/typescript.test.ts.snap]: missing documentation here too

File: packages/compiler/src/__tests__/__snapshots__/typescript.test.ts.snap [hunk 13]
```diff
@@ -1219,21 +1020,27 @@ This is equivalent to setting \`className=\\"pt-active\\"\`.",
...
+, the button will display a centered loading spinner instead of its contents.
 The width of the button is not affected by the value of this prop.",
-          "metadata": Object {},
+          "metadata": {},
         },
         "fileName": "src/__tests__/__fixtures__/interfaces.ts",
-        "flags": Object {
-          "isDeprecated": true,
+        "flags": {
+          "isDeprecated": "",
```
Comment (adidahiya) [inline packages/compiler/src/__tests__/__snapshots__/typescript.test.ts.snap]: this is a potential regression

File: packages/compiler/src/plugins/typescript/visitor.ts [hunk 14]
```diff
@@ -33,14 +33,15 @@ import {
...
 import { relative } from "path";
 import {
+    Comment,
     DeclarationReflection,
     ParameterReflection,
     ProjectReflection,
     Reflection,
     ReflectionKind,
     SignatureReflection,
+    // UnionType,
```
Comment (adidahiya) [inline packages/compiler/src/plugins/typescript/visitor.ts]: left some commented code in this file which needs to be deleted

File: .circleci/config.yml (hunk 5 above)
Comment (adidahiya) [reply .circleci/config.yml]: extracted here: https://github.com/palantir/documentalist/pull/236

File: .circleci/config.yml (hunk 4 above)
Comment (adidahiya) [reply .circleci/config.yml]: extracted here: https://github.com/palantir/documentalist/pull/236

**Comment by adidahiya** [comment]
Build preview link for commit "Merge remote-tracking branch 'origin/develop' into ad/upgrade-typescript": <strong><a href="https://output.circle-artifacts.com/output/job/39cf8418-8341-4581-b022-463967823a79/artifacts/0/packages/docs/dist/index.html">documentation</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

File: packages/compiler/src/__tests__/__snapshots__/typescript.test.ts.snap (hunk 8 above)
Comment (adidahiya) [reply packages/compiler/src/__tests__/__snapshots__/typescript.test.ts.snap]: found out that this is due to a [breaking change in v0.23.0](https://typedoc.org/guides/changelog/#v0.23.0-(2022-06-26)):

> `reflection.defaultValue` is no longer set for enum members. The same information is available on `reflection.type` with more precision.

will fix!

File: packages/compiler/src/__tests__/__snapshots__/typescript.test.ts.snap (hunk 6 above)
Comment (adidahiya) [reply packages/compiler/src/__tests__/__snapshots__/typescript.test.ts.snap:431]: Note that this was actually sort of a _regression_ in https://github.com/palantir/documentalist/pull/144, so removing `Food` here is a good thing.

File: packages/compiler/src/__tests__/__snapshots__/typescript.test.ts.snap (hunk 8 above)
Comment (adidahiya) [reply packages/compiler/src/__tests__/__snapshots__/typescript.test.ts.snap]: ✅ fixed

**Comment by adidahiya** [comment]
Build preview link for commit "implement multi-ts-project support": <strong><a href="https://output.circle-artifacts.com/output/job/0a4b46f3-50a2-43a8-9f3e-a1153e3b87a8/artifacts/0/packages/docs/dist/index.html">documentation</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

**Comment by adidahiya** [comment]
Build preview link for commit "Merge remote-tracking branch 'origin/develop' into ad/upgrade-typescript": <strong><a href="https://output.circle-artifacts.com/output/job/a1c8f6d2-531e-4bf5-a34c-e4b64a531051/artifacts/0/packages/docs/dist/index.html">documentation</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

**Comment by adidahiya** [comment]
Build preview link for commit "fix lint": <strong><a href="https://output.circle-artifacts.com/output/job/1a857ac1-4904-400b-971b-f2b8a34610ba/artifacts/0/packages/docs/dist/index.html">documentation</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

File: packages/compiler/src/__tests__/__snapshots__/typescript.test.ts.snap (hunk 6 above)
Comment (adidahiya) [reply packages/compiler/src/__tests__/__snapshots__/typescript.test.ts.snap:431]: also, we can get this back if we add https://www.npmjs.com/package/typedoc-plugin-missing-exports. so we will have options after this upgrade.

**Comment by adidahiya** [comment]
Build preview link for commit "remove missing exports plugin": <strong><a href="https://output.circle-artifacts.com/output/job/7e09078c-566b-4d74-839b-fc072788cdb0/artifacts/0/packages/docs/dist/index.html">documentation</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

File: packages/compiler/cli.js [hunk 15]
```diff
@@ -46,6 +47,10 @@ const argv = yargs
         desc: "use KssPlugin for .(css|less|scss) files",
         type: "boolean",
     })
+    .option("out", {
```
Comment (adidahiya) [inline packages/compiler/cli.js:50]: adding this option allows us to log things to the console and still use the CLI to output a JSON file (previously, we were just using stdout in the docs package of this repo)

File: packages/compiler/src/typedoc-plugin-missing-exports.d.ts [hunk 16]
```diff
@@ -0,0 +1,4 @@
+declare module "typedoc-plugin-missing-exports" {
```
Comment (adidahiya) [inline packages/compiler/src/typedoc-plugin-missing-exports.d.ts]: I should remove this file until we're ready to use this library

**Comment by adidahiya** [comment]
Build preview link for commit "fix lint again": <strong><a href="https://output.circle-artifacts.com/output/job/ed2badc5-0370-44ef-a31c-23fa638cc174/artifacts/0/packages/docs/dist/index.html">documentation</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

File: packages/compiler/src/__tests__/__snapshots__/typescript.test.ts.snap (hunk 7 above)
Comment (adidahiya) [reply packages/compiler/src/__tests__/__snapshots__/typescript.test.ts.snap:221]: actually, it looks like we still have the get & set documentation, but this redundant compound "documentation" for get + set together has been removed. that's fine ✅

**Comment by adidahiya** [comment]
Build preview link for commit "Handle @deprecated tag better": <strong><a href="https://output.circle-artifacts.com/output/job/40a8cfd7-7ff0-4dc2-9998-eb14b8e6194c/artifacts/0/packages/docs/dist/index.html">documentation</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

File: packages/compiler/src/__tests__/__snapshots__/typescript.test.ts.snap (hunk 10 above)
Comment (adidahiya) [reply packages/compiler/src/__tests__/__snapshots__/typescript.test.ts.snap]: fixed now ✅

File: packages/compiler/src/__tests__/__snapshots__/typescript.test.ts.snap (hunk 12 above)
Comment (adidahiya) [reply packages/compiler/src/__tests__/__snapshots__/typescript.test.ts.snap]: fixed now ✅

**Comment by adidahiya** [comment]
Build preview link for commit "Fix interface property signatures": <strong><a href="https://output.circle-artifacts.com/output/job/d5283a8e-79ed-48f2-b764-b186addcd792/artifacts/0/packages/docs/dist/index.html">documentation</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

**Comment by adidahiya** [comment]
Build preview link for commit "fix lint": <strong><a href="https://output.circle-artifacts.com/output/job/64960621-3a56-4355-8f97-35b5a8a96d5d/artifacts/0/packages/docs/dist/index.html">documentation</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

File: packages/compiler/src/__tests__/__snapshots__/typescript.test.ts.snap (hunk 11 above)
Comment (adidahiya) [reply packages/compiler/src/__tests__/__snapshots__/typescript.test.ts.snap:879]: I believe this is a bug in typedoc, filed this issue: https://github.com/TypeStrong/typedoc/issues/2414

But I don't think we need to block on this; this is an uncommonly used feature.

**Comment by adidahiya** [comment]
Build preview link for commit "update more snapshots": <strong><a href="https://output.circle-artifacts.com/output/job/63a7c695-9db0-4087-b7d9-99497dc1cb85/artifacts/0/packages/docs/dist/index.html">documentation</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

**Comment by adidahiya** [comment]
Build preview link for commit "remove extraneous type definitions": <strong><a href="https://output.circle-artifacts.com/output/job/c5f5ceaa-6f96-477c-9ae8-440264b3622f/artifacts/0/packages/docs/dist/index.html">documentation</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

File: packages/compiler/package.json [hunk 17]
```diff
@@ -22,25 +22,26 @@
...
-        "yargs": "^17.7.1"
+        "kss": "^3.0.1",
+        "marked": "^4.0.12",
+        "tsconfig-resolver": "^3.0.1",
+        "typedoc": "~0.25.2",
+        "yargs": "^17.4.0"
     },
     "devDependencies": {
-        "@types/glob": "^8.0.1",
+        "@types/glob": "^7.2.0",
```
Comment (styu) [inline packages/compiler/package.json]: ooc why did you have to downgrade?

File: packages/compiler/src/plugins/typescript/visitor.ts [hunk 18]
```diff
@@ -199,56 +199,61 @@ export class Visitor {
...
 }
 
-function getSourceUrl({ sources = [] }: Reflection): string | undefined {
-    const source = sources[0];
-    return source && source.url;
+function getSourceUrl(reflection: Reflection): string | undefined {
+    if (reflection.isDeclaration() || isSignatureReflection(reflection)) {
+        if (reflection.sources !== undefined) {
+            const source = reflection.sources[0];
+            return source && source.url;
```
Comment (styu) [inline packages/compiler/src/plugins/typescript/visitor.ts]: could/should we do this now?
```suggestion
            return source?.url;
```

**Review by styu — APPROVED** [review]

File: packages/compiler/package.json (hunk 17 above)
Comment (adidahiya) [reply packages/compiler/package.json]: hmm, I think this was a bad rebase/merge since this branch was originally started a long time ago. good catch -- will fix

**Comment by adidahiya** [comment]
Build preview link for commit "address minor review comments": <strong><a href="https://output.circle-artifacts.com/output/job/3293c8b5-e365-4030-8ee4-2b31b821c461/artifacts/0/packages/docs/dist/index.html">documentation</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

**Comment by adidahiya** [comment]
Build preview link for commit "don't capitalize TypeScript in public API": <strong><a href="https://output.circle-artifacts.com/output/job/b527dc08-d9f0-44cd-a842-3999e839a735/artifacts/0/packages/docs/dist/index.html">documentation</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>
